examples/simple_bind_authentication/src/main.rs

```diff
@@ -0,0 +1,132 @@
+#[macro_use]
+extern crate clap;
+
+extern crate openldap;
+
+use openldap::errors::*;
+use openldap::*;
+use std::ptr;
+
+#[derive(Clap)]
+#[clap(
+    name = "LDAP simple_bind_authentication with start_tls authentication example",
+    author = "Mathias Myrland <jedimemo@gmail.com>",
+    version = "0.1.0"
+)]
+struct AuthOpts {
+    #[clap(short = "u")]
+    user: String,
+
+    #[clap(short = "p")]
+    password: String,
+}
+
+fn ldap_with_start_tls(ldap_uri: &str) -> Result<RustLDAP, LDAPError> {
+    let ldap = RustLDAP::new(ldap_uri).unwrap();
+
+    ldap.set_option(
+        codes::options::LDAP_OPT_PROTOCOL_VERSION,
+        &codes::versions::LDAP_VERSION3,
+    );
+
+    // WARNING: Normally you would want to verify the server certificate to avoid
+    // man in the middle attacks, but for this testing scenario we're using a
+    // generated self signed certificate from the docker container.
+    //
+    // To set up certificate validation, use the LDAP_OPT_X_TLS_CACERT* options
+    ldap.set_option(
+        codes::options::LDAP_OPT_X_TLS_REQUIRE_CERT,
+        &codes::options::LDAP_OPT_X_TLS_NEVER,
+    );
+
+    ldap.set_option(openldap::codes::options::LDAP_OPT_X_TLS_NEWCTX, &0);
+
+    ldap.start_tls(None, None)?;
+
+    Ok(ldap)
+}
+
+fn do_simple_bind(
+    ldap: &RustLDAP,
+    ldap_manager_user: &str,
+    ldap_manager_pass: &str,
+) -> Result<(), LDAPError> {
+    let bind_result = ldap.simple_bind(ldap_manager_user, ldap_manager_pass)?;
+
+    match bind_result {
+        v if v == openldap::codes::results::LDAP_SUCCESS => Ok(()),
+        _ => Err(LDAPError::from(String::from(
+            "Authentication with simple bind failed",
+        ))),
+    }
+}
+
+fn ldap_dn_lookup(ldap: &RustLDAP, who: &str) -> Result<String, LDAPError> {
+    // Show all DNs matching the description "Human"
+    // ldap_search is a powerful query language, look at
+    // https://confluence.atlassian.com/kb/how-to-write-ldap-search-filters-792496933.html
+    // for an overview
+    //
+    // This particular filter allows the user to sign in with either
+    // uid or email
+    let filter = format!("(|(uid={})(mail={}))", who, who);
```

Comment:
I don’t think the C LDAP library provided this functionality, and since this is a thin wrapper around it it was never present. The C library likely assumed that you escaped your queries yourself, as I did when I was using it for the project the fork was made for.

That being said, I do not think it’s a bad idea to add that ability if there is a standardized way to perform it and it is not required to be done if you can prove that your queries are otherwise ‘safe’.